src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -1845,6 +1847,29 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
                 );
                 throw SemanticAbort();
             }
+            std::string obj = "";
+            if (ASR::is_a<ASR::List_t>(*value_type)) {
+                target_type = ASR::down_cast<ASR::List_t>(target_type)->m_type;
+                value_type = ASR::down_cast<ASR::List_t>(value_type)->m_type;
+                obj = "list";
+            } else if (ASR::is_a<ASR::Set_t>(*value_type)) {
+                target_type = ASR::down_cast<ASR::Set_t>(target_type)->m_type;
+                value_type = ASR::down_cast<ASR::Set_t>(value_type)->m_type;
+                obj = "set";
+            }
```

Comment:
Since you only use `obj` in the error below, you should move that inside the `if` statement.

I think we need to move the `List` downcast inside the `ASRUtils::check_equal_type`, not doing it before hand.